### ISSUE TITLE
Warnings() in read.ctd.odf function::oce

### ISSUE THREAD

**Issue by torkjot** [dankelley/oce#1433]
[DCOR2017053sample.txt](https://github.com/dankelley/oce/files/2145688/DCOR2017053sample.txt)

Attached: ~5 lines of sample data file.

### Short summary of the problem:

I have encountered warnings() from the `read.ctd.odf` function within oce.
```R
Error in findInHeader("EVENT_QUALIFIER", lines) : 
  (converted from warning) NAs introduced by coercion.
```

### What you did:
```R
setwd("~/input/")
options(warn=3)
lst <- list.files(pattern="DCOR2017.*.ODF")
sectionlst <- list()
for (i in seq_along(lst)){
    read_in_file <- read.ctd.odf(lst[i],debug=3)
    sectionlst[[i]] <- read_in_file
} 
```
### What you expected to happen:

I expected no warnings.

### What happened
```R
Error in findInHeader("EVENT_QUALIFIER", lines) : 
  (converted from warning) NAs introduced by coercion
```
Please let me know if more information is required - Tor

**Comment by dankelley** [comment]
Hm, this is interesting. I cannot reproduce the problem on my machine
```R
> library(oce)
Loading required package: gsw
Loading required package: testthat
> d<-read.oce("~/Dropbox/DCOR2017053sample.odf")
```
From our f2f discussion, this seems related to time, so I'm looking at that and get
```R
> d[["time"]]
[1] "2017-04-22 07:14:05 UTC" "2017-04-22 07:14:12 UTC" "2017-04-22 07:14:27 UTC" "2017-04-22 07:14:30 UTC"
[5] "2017-04-22 07:14:31 UTC" "2017-04-22 07:14:35 UTC" "2017-04-22 07:14:35 UTC" "2017-04-22 07:14:38 UTC"
[9] "2017-04-22 07:14:41 UTC"
```

Tor, can you try the above and note your results in a comment? That may help me to find the problem.

**Comment by torkjot** [comment]
When I try your note I get:
> d<-read.oce("~/Desktop/DCOR2017053sample.odf")
> d[["time"]]
[1] "2017-04-22 07:14:05 UTC" "2017-04-22 07:14:12 UTC" "2017-04-22 07:14:27 UTC"
[4] "2017-04-22 07:14:30 UTC" "2017-04-22 07:14:31 UTC" "2017-04-22 07:14:35 UTC"
[7] "2017-04-22 07:14:35 UTC" "2017-04-22 07:14:38 UTC" "2017-04-22 07:14:41 UTC"

If it is related to packaging, my current setup is:
install.packages("~/Desktop/osx/oce_0.9-24.tgz", repos = NULL, type = .Platform$pkgType)
install.packages("~/Desktop/osx/ocedata_0.1.4.tgz", repos = NULL, type = .Platform$pkgType)
install.packages("gsw")
install.packages("Rcpp")

library(Rcpp)
library(testthat)
library(gsw)
library(oce)
library(ocedata)

**Comment by dankelley** [comment]
Can you put the results from
```R
str(d@metadata)
```
NOTE: you ought to format your output as in the screenshot for what I wrote above.

![screen shot 2018-06-28 at 11 51 28 am](https://user-images.githubusercontent.com/99469/42042049-a3886bee-7ac9-11e8-89a7-dca22d7e21ea.png)

**Comment by torkjot** [comment]
```R
str(d@metadata)
List of 32
 $ flags               :List of 1
  ..$ overall: num [1:9] 0 0 0 0 0 0 0 0 0
 $ pressureType        : chr "sea"
 $ deploymentType      : chr "profile"
 $ waterDepth          : num 2267
 $ startTime           : POSIXct[1:1], format: "2017-04-22 07:08:07"
 $ eventNumber         : num 53
 $ eventQualifier      : chr "1 of 2."
 $ filename            : chr "/Users/Tjensvoll/Desktop/DCOR2017053sample.odf"
 $ serialNumber        : num NA
 $ ship                : chr "Coriolis II"
 $ cruise              : chr "Spring AZMP Scotian Shelf"
 $ station             : num 53
 $ scientist           : chr "Andrew Cogswell"
 $ units               :List of 27
  ..$ scan          :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr ""
  ..$ time          :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr ""
  ..$ pressure      :List of 2
  .. ..$ unit :  expression(dbar)
  .. ..$ scale: chr ""
  ..$ temperature   :List of 2
  .. ..$ unit :  expression(degree * C)
  .. ..$ scale: chr "IPTS-68"
  ..$ conductivity  :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr ""
  ..$ temperature2  :List of 2
  .. ..$ unit :  expression(degree * C)
  .. ..$ scale: chr "IPTS-68"
  ..$ conductivity2 :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr ""
  ..$ altimeter     :List of 2
  .. ..$ unit :  expression(m)
  .. ..$ scale: chr ""
  ..$ par           :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr ""
  ..$ oxygenVoltage :List of 2
  .. ..$ unit :  expression(V)
  .. ..$ scale: chr ""
  ..$ oxygenVoltage2:List of 2
  .. ..$ unit :  expression(V)
  .. ..$ scale: chr ""
  ..$ fluorometer   :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr ""
  ..$ fluorometer2  :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr ""
  ..$ pH            :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr ""
  ..$ unknown       :List of 2
  .. ..$ unit :  expression(1/m/sr)
  .. ..$ scale: chr ""
  ..$ latitude      :List of 2
  .. ..$ unit :  expression(degree)
  .. ..$ scale: chr ""
  ..$ longitude     :List of 2
  .. ..$ unit :  expression(degree)
  .. ..$ scale: chr ""
  ..$ salinity      :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr "PSS-78"
  ..$ salinity2     :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr "PSS-78"
  ..$ theta         :List of 2
  .. ..$ unit :  expression(degree * C)
  .. ..$ scale: chr "IPTS-68"
  ..$ theta2        :List of 2
  .. ..$ unit :  expression(degree * C)
  .. ..$ scale: chr "IPTS-68"
  ..$ sigmaTheta    :List of 2
  .. ..$ unit :  expression(kg/m^3)
  .. ..$ scale: chr ""
  ..$ sigmaTheta2   :List of 2
  .. ..$ unit :  expression(kg/m^3)
  .. ..$ scale: chr ""
  ..$ oxygen        :List of 2
  .. ..$ unit :  expression(ml/l)
  .. ..$ scale: chr ""
  ..$ oxygen2       :List of 2
  .. ..$ unit :  expression(ml/l)
  .. ..$ scale: chr ""
  ..$ scan2         :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr ""
  ..$ overallFlag   :List of 2
  .. ..$ unit :  expression()
  .. ..$ scale: chr ""
 $ longitude           : num -61.5
 $ latitude            : num 42.6
 $ dataNamesOriginal   :List of 26
  ..$ scan          : chr "CNTR_01"
  ..$ time          : chr "SYTM_01"
  ..$ pressure      : chr "PRES_01"
  ..$ temperature   : chr "TEMP_01"
  ..$ conductivity  : chr "CRAT_01"
  ..$ temperature2  : chr "TEMP_02"
  ..$ conductivity2 : chr "CRAT_02"
  ..$ altimeter     : chr "ALTB_01"
  ..$ par           : chr "PSAR_01"
  ..$ oxygenVoltage : chr "OXYV_01"
  ..$ oxygenVoltage2: chr "OXYV_02"
  ..$ fluorometer   : chr "FLOR_01"
  ..$ fluorometer2  : chr "FLOR_02"
  ..$ pH            : chr "PHPH_01"
  ..$ unknown       : chr "UNKN_01"
  ..$ latitude      : chr "LATD_01"
  ..$ longitude     : chr "LOND_01"
  ..$ salinity      : chr "PSAL_01"
  ..$ salinity2     : chr "PSAL_02"
  ..$ theta         : chr "POTM_01"
  ..$ theta2        : chr "POTM_02"
  ..$ sigmaTheta    : chr "SIGP_01"
  ..$ sigmaTheta2   : chr "SIGP_02"
  ..$ oxygen        : chr "DOXY_01"
  ..$ oxygen2       : chr "DOXY_02"
  ..$ scan2         : chr "CNTR_02"
 $ depthOffBottom      : num -53
 $ initialLatitude     : num 42.6
 $ initialLongitude    : num -61.5
 $ endLatitude         : num -99.9
 $ endLongitude        : num -1000
 $ samplingInterval    : num 0
 $ institute           : chr "BIO"
 $ countryInstituteCode: num 1810
 $ cruiseNumber        : chr "COR2017001"
 $ depthMin            : num 1
 $ depthMax            : num 2320
 $ sounding            : num 2267
 $ type                : chr "SBE"
 $ model               : chr "SBE 9"
 $ sampleInterval      : logi NA
```

Apologies for previous formatting.

**Comment by dankelley** [comment]
I think we need to see where the error is. If you're willing, please download the file `R/odf.R` from the oce source on github (in the "develop" branch, not the "master" branch) and then do
```R
source("odf.R")
options(warn=3)
d <- read.odf("~/Desktop/DCOR2017053sample.odf")
```
which should give an error message, then do
```R
traceback()
```
which should give the line number of the error. If you paste the full output from `traceback()` here, I may be able to see the problem.  (If this is a time thing then it could be an issue of our two computers being set up for different locales. But I want `read.odf()` to work on all locales, of course!)

**Comment by torkjot** [comment]
I copied the ```R odf.R ```  from: https://github.com/dankelley/oce/blob/develop/R/odf.R to a new script, as I couldn't figure out how to download it and then ran the following.
```R
> source("odf.R")
> options(warn=3)
> d <- read.odf("~/Desktop/DCOR2017053sample.odf")
> traceback()
2: file(filename, "r", encoding = encoding)
1: source("odf.R")
```
There did not appear to be an error message from ```R read.odf ```.

**Comment by dankelley** [comment]
Sorry, I didn't realize which function you were using. Please try also sourcing `read.ctd.odf.R` and running the test with that function.

**Comment by torkjot** [comment]
Using ```R oce/ctd.odf.R ``` :
```R
> setwd("~/Desktop/")
> source("ctd.odf.R")
> options(warn=3)
> d <- read.ctd.odf("~/Desktop/DCOR2017053sample.odf")
> traceback()
6: doWithOneRestart(return(expr), restart)
5: withOneRestart(expr, restarts[[1L]])
4: withRestarts({
       .Internal(.signalCondition(simpleWarning(msg, call), msg, 
           call))
       .Internal(.dfltWarn(msg, call))
   }, muffleWarning = function() NULL)
3: .signalSimpleWarning("cannot open file 'ctd.odf.R': No such file or directory", 
       quote(file(filename, "r", encoding = encoding)))
2: file(filename, "r", encoding = encoding)
1: source("ctd.odf.R")
```

**Comment by dankelley** [comment]
Oh, my. I have a meeting in 1/2 hour so cannot ask you to drop by, but that is likely the best way to make progress.

Q: when you built up oce, did you use `install_github` or did you download the source and install it in the more basic way?  I think the way to make progress is for you to download it, build it, and then run your test. I think that then `traceback()` will give us the line number.

The next step (which you may not be comfortable doing) is to uncomment the `message` and `print` calls in `findInHeader` in `R/odf.R`, and rerun. This should give us the line that is problematic. Also, the Rstudio could be set up to open a browser, when there is an error. That way, we can use interactive commands like `print(lines[j])` and we'll see the problem line.

This is the king of thing that's way faster in person.

**Comment by richardsc** [comment]
@torkjot I assume you're on Windows? The output of a `sessionInfo()` might be helpful.

I can try things on my Windows machine a bit later today.

**Comment by torkjot** [comment]
@richardsc I am on Mac.
@dankelley I will give your comments a try.

**Comment by torkjot** [comment]
When I built up oce, I downloaded the osx folder of the Pre-compiled development version from: http://dankelley.github.io/oce/installation.html , which is what I believe is the more basic way. I can re-try with the source-code version if that is suggested. 
I found some difficulties when building oce as ```R testthat , gsw , and Rcpp ``` required non-development versions. I am not sure if this is where my fault is. I am not running the development version of ```R gsw ``` for instance as errors occurred when I attempted to call it from library.
When attempting to use the ```R gsw ``` development version the following error occurs:
```R
Error: package or namespace load failed for ‘gsw’:
 package ‘gsw’ was installed by an R version with different internals; it needs to be reinstalled for use with this R version
```
The CRAN version of ```R gsw ``` can be called without error. I am unsure if this helps.
I am attempting the latter part of your comment now.

**Comment by torkjot** [comment]
I have now just tried with a different data file to see if there is a difference. I am re-attaching a new file which I now know for sure causes a problem.
[DCOR2017054 copy.txt](https://github.com/dankelley/oce/files/2146498/DCOR2017054.copy.txt)
I have checked this file with the initial method mentioned:

```R
> setwd("~/Desktop/")
> source("ctd.odf.R")
> options(warn=3)
> d <- read.ctd.odf("~/Desktop/DCOR2017053sample.odf")
> traceback()
No traceback available 
> d2 <- read.ctd.odf("~/Desktop/DCOR2017054 copy.odf")
 Show Traceback
 
 Rerun with Debug
 Error in findInHeader("EVENT_QUALIFIER", lines) : 
  (converted from warning) NAs introduced by coercion 

> traceback()
7: doWithOneRestart(return(expr), restart)
6: withOneRestart(expr, restarts[[1L]])
5: withRestarts({
       .Internal(.signalCondition(simpleWarning(msg, call), msg, 
           call))
       .Internal(.dfltWarn(msg, call))
   }, muffleWarning = function() NULL)
4: .signalSimpleWarning("NAs introduced by coercion", quote(findInHeader("EVENT_QUALIFIER", 
       lines)))
3: findInHeader("EVENT_QUALIFIER", lines)
2: read.odf(file = file, columns = columns, debug = debug - 1) at ctd.odf.R#23
1: read.ctd.odf("~/Desktop/DCOR2017054 copy.odf")
```

**Comment by richardsc** [comment]
For what it's worth I get the same warning. It doesn't seem to interfere with actually reading the file.

My `sessionInfo()`:
```
R version 3.4.2 (2017-09-28)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: OS X El Capitan 10.11.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] oce_0.9-24     gsw_1.0-5      testthat_1.0.2

loaded via a namespace (and not attached):
[1] compiler_3.4.2 magrittr_1.5   R6_2.2.1       tools_3.4.2    yaml_2.1.14    crayon_1.3.2  
```

Haven't dug in yet to see if I can get the line number.

**Comment by dankelley** [comment]
Good. I get the error now, and I'll fix it tomorrow.
```R
> d<-read.oce("~/Dropbox/DCOR2017054.odf")
Warning message:
In findInHeader("EVENT_QUALIFIER", lines) : NAs introduced by coercion
> 
```

One thing I see is that this file has two lines that contain "`EVENT_QUALIFIER`", viz lines 15 and 16
```
   EVENT_QUALIFIER1='**',
   EVENT_QUALIFIER2='DN',
```

PS. This is one of these files that start each line with a blank, which seems like maybe it's a new scheme or something.

**Comment by richardsc** [comment]
Here's what I get:
```
> traceback()
7: doWithOneRestart(return(expr), restart)
6: withOneRestart(expr, restarts[[1L]])
5: withRestarts({
       .Internal(.signalCondition(simpleWarning(msg, call), msg, 
           call))
       .Internal(.dfltWarn(msg, call))
   }, muffleWarning = function() NULL)
4: .signalSimpleWarning("NAs introduced by coercion", quote(findInHeader("EVENT_QUALIFIER", 
       lines))) at odf.R#265
3: findInHeader("EVENT_QUALIFIER", lines) at odf.R#1133
2: read.odf(file = file, columns = columns, debug = debug - 1) at ctd.odf.R#23
1: read.ctd.odf("~/Downloads/DCOR2017054.copy.txt")
```

**Comment by dankelley** [comment]
Yes, I get it directly with
```R
> lines<-readLines("~/Dropbox/DCOR2017054.odf",encoding="latin1")
> oce:::findInHeader("EVENT_QUALIFIER", lines)
[1] NA
Warning message:
In oce:::findInHeader("EVENT_QUALIFIER", lines) :
  NAs introduced by coercion
```

**Comment by dankelley** [comment]
I see that
```R
> options(warn=3)
> oce:::findInHeader("EVENT_QUALIFIER", lines)
Error in oce:::findInHeader("EVENT_QUALIFIER", lines) : 
  (converted from warning) NAs introduced by coercion
> traceback()
5: doWithOneRestart(return(expr), restart)
4: withOneRestart(expr, restarts[[1L]])
3: withRestarts({
       .Internal(.signalCondition(simpleWarning(msg, call), msg, 
           call))
       .Internal(.dfltWarn(msg, call))
   }, muffleWarning = function() NULL)
2: .signalSimpleWarning("NAs introduced by coercion", quote(oce:::findInHeader("EVENT_QUALIFIER", 
       lines))) at odf.R#265
1: oce:::findInHeader("EVENT_QUALIFIER", lines)
```
and that tells me the problem.

**Comment by dankelley** [comment]
I think this ought to work now in the "develop" branch, commit e3cbd7e12c286974fa7dcbe54425c9298fa0afa9 (to Tor: if you click the link to the left, you'll see the code changes). Below is what I get now, i.e. no warnings.
```R
> library(oce)
> d<-read.oce("~/Dropbox/DCOR2017054.odf")
> d<-read.ctd.odf("~/Dropbox/DCOR2017054.odf")
> d<-read.odf("~/Dropbox/DCOR2017054.odf")
```

I ran my odf test suite, which reads 973 .odf files that I have on my machine, and that works OK, although all it really does is to read the files and summarize them.

So, this is ready for tests.

**NOTE** oce is not actually *using* these `EVENT_QUALIFIER` items. This is because I don't have much idea what they are permitted to contain. In this case, there are two qualifiers, and the first is `"**"`, the meaning of which escapes me, and the second is `"DN"`, which likely means a downcast. But I am only reading *one* of these two items, namely the first. (If someone really needs oce to handle event-qualifiers, that would be a separate issue, I think ... after all, we've had code to read just the one of them for years now, and I only realized in the past half hour that files can contain two!)

Tor, the rule on oce issues is that it's up to the reporter to decide whether the issue has been addressed. No rush -- do some tests at your leisure. And thanks a lot for the report!

D.
